### ISSUE TITLE
Docker image enhancements

### ISSUE THREAD

**Issue by stenmigerode** [inbo/reporting-rshiny-grofwildjacht#313]
# TL;DR

Incorporate docker multi-stage builds and improve Dockerfile for generating smaller docker images. And externalize the data as described in #280

Suggestions:

* Only install apt packages that are needed
* Install apt packages that are needed to build the application in the build stage
* Start from a fresh image in the last stage and copy the built application
* Reduce R package sizes? Is it possible?
* Merge run commands on one line, especially if you are deleting temp files afterwards (not needed of course if you could do this in the build stage)

# Detailed description

The docker image of the the grofwildjacht is quite big and there are a couple of improvements that could reduce the image size and build time.

## Externalize data

First possible improvement is already discussed in #280 by externalizing the data. The data files could be mounted using docker volumes into the docker image. That way new data should not trigger new docker image builds.

## Multi-stage docker builds

The second improvement is to use multi-stage builds in Docker: https://docs.docker.com/develop/develop-images/multistage-build/

Every single command in a Dockerfile adds another layer in the docker image. You can check these layers in docker images with the following command:

```bash
docker history openanalytics/wildapp
```

### Current situation

![afbeelding](https://user-images.githubusercontent.com/27133403/170252506-f14c1891-e4f8-46a3-a410-cb567fe309fd.png)

We can see the different build steps in the above screenshot.

We can see that the following command will add over 700MB to the docker image:
https://github.com/inbo/reporting-rshiny-grofwildjacht/blob/66ff3e447dd5c48ea8bcb29c7db66f84aea787c3/Dockerfile#L6-L21

A couple of apt packages are installed here: how many of these are required for the build only? And how many of these are really needed? For example: is sudo really needed for the application? Normally you would perform the docker build steps with the root user and switch to a non-root user at the end of the installation.

Also for the following snippet:
https://github.com/inbo/reporting-rshiny-grofwildjacht/blob/66ff3e447dd5c48ea8bcb29c7db66f84aea787c3/Dockerfile#L27-L48

* Is it possible to reduce the R package sizes (e.g. if these packages include the source code, can you switch to the binaries only?)
* Is it possible to add this step in the build stage and copy the built packages together with the built application to the final stage ?

## Minimizing the number of commands in the Dockerfile

Aside from multi-stage builds, the last image should contain the least amount of RUN commands. Because each RUN command will add a layer. That also means that if you add one file in 1 command, and delete it again in another command, the size of the image will not be reduced.

We can see this here:

![afbeelding](https://user-images.githubusercontent.com/27133403/170255778-06d24b16-d966-4a78-ae55-7b06db596595.png)

https://github.com/inbo/reporting-rshiny-grofwildjacht/blob/66ff3e447dd5c48ea8bcb29c7db66f84aea787c3/Dockerfile#L52-L55

The copy of the tar.gz file adds 59MB, the installation adds 118MB, the removal of the tar.gz does nothing: it does not reduce the image size since its added on a different layer.

By merging these 3 RUN commands in 1 RUN command, the remove command will have it's intented effect: it will reduce the size of the image.

# Incentive for this improvement
The build times are quite long (and use cloud resources). Using multi-stage build and caches can drastically improve the build times.

Currently the application is running on quite a big AWS server instance (4vCPU, 16 GB). Because the WBE private data requires a second instance of the ShinyProxy, anything that we can do to scale down the server requirements will reduce costs 2 times.
